feat: Draft through API for Cross-Device Support

jderecho opened the pull request:
# Pull Request Template

## Description

Fixes https://github.com/chatwoot/chatwoot/issues/6890

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

pranavrajs commented on app/controllers/api/v1/accounts/conversations_controller.rb:
```diff
@@ -80,6 +80,12 @@ def toggle_typing_status
   end
 
+  def update_draft_message
+    @conversation.draft_message_content = params.permit(:message)[:message]
```
Saving it as an attribute to conversation is not ideal. There are multiple people looking at the same conversation. Showing a draft saved by one of the users to another user is not correct.

pranavrajs replied:
1/ Saving the drafts in a conversation is not ideal. As there are multiple users looking at the same conversation, exposing a draft written by one user to another user is not a great experience. 

2/ The draft message is short-living item, it shouldn't be stored in the database instead it can be saved in Redis storage with key to identify the user, the account and the conversation. The key value pair can be deleted once the message is sent.

pranavrajs commented on app/controllers/api/v1/accounts/conversations_controller.rb:
```diff
@@ -80,6 +80,12 @@ def toggle_typing_status
   end
 
+  def update_draft_message
+    @conversation.draft_message_content = params.permit(:message)[:message]
```
1/ Saving the drafts in a conversation is not ideal. As there are multiple users looking at the same conversation, exposing a draft written by one user to another user is not a great experience.

2/ The draft message is short-living item, it shouldn't be stored in the database instead it can be saved in Redis storage with key to identify the user, the account and the conversation. The key value pair can be deleted once the message is sent.

pranavrajs replied:
3/ Let us consider draftMessages as a separate entity rather than tying it to conversation. This would give them a separate namespace for APIS. `/accounts/<account-id>/conversations/<conversation-id>/draft-messages` or something similar would work.

pranavrajs commented on app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue:
```diff
@@ -629,7 +630,7 @@ export default {
         const key = `draft-${this.conversationIdByRoute}-${this.replyType}`;
         const messageFromStore =
-          this.$store.getters['draftMessages/get'](key) || '';
+          this.$store.getters['draftMessages/get'](key) || this.currentChat.draft_message_content || '';
```
Same comment as above. We should not be polluting the conversation namespace rather consider the draftMessage as an entity by itself.

pranavrajs commented on app/javascript/dashboard/api/inbox/conversation.js:
```diff
@@ -112,6 +112,12 @@ class ConversationApi extends ApiClient {
     return axios.post(`${this.url}/${conversationId}/transcript`, { email });
   }
 
+  updateDraftMessage({ conversationId, message }) {
```
As mentioned above, let us move draftMessages from the conversation namespace to its own namespace.

pranavrajs commented on app/javascript/dashboard/store/modules/conversations/actions.js:
```diff
@@ -405,6 +405,21 @@ const actions = {
     }
   },
 
+  updateDraftMessage: async (
```
We have a store called draftMessages. Let us use that. 

https://github.com/chatwoot/chatwoot/blob/develop/app/javascript/dashboard/store/modules/draftMessages.js

pranavrajs reviewed (changes requested):
Thanks for the PR. The implementation needs a bit of work. Let me know if you have any questions.

jderecho commented:
hi @pranavrajs I refactored the code to use Redis caching, let me know if I implemented it correctly and if there are other things to consider. Thanks

pranavrajs commented:
@jderecho Thanks for addressing the comments. On a quick look, this looks fine. Let me test it out properly and complete the review.

sojan-official commented:
Note: We will be decoupling the frontend changes in this PR separately.

sojan-official commented:
Note: Follow-up enhancement -> add user ids to the drafts to differentiate drafts between users
